internal/pkg/runner/runner.go

```diff
@@ -0,0 +1,80 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package runner
+
+import (
+	"context"
+	"sync"
+	"time"
+)
+
+type RunnerFunc func(context.Context) error
+
+type Runner struct {
+	fn RunnerFunc
+	cn context.CancelFunc
+
+	mx   sync.Mutex
+	done chan struct{}
+	err  error
+}
+
+func (r *Runner) Stop() {
+	r.mx.Lock()
+	if r.cn != nil {
+		r.cn()
+		r.cn = nil
+	}
+	r.mx.Unlock()
+}
+
+func (r *Runner) Err() error {
+	return r.err
+}
+
+func (r *Runner) Done() <-chan struct{} {
+	return r.done
+}
+
+func (r *Runner) DoneWithTimeout(to time.Duration) <-chan struct{} {
+	done := make(chan struct{})
+
+	go func() {
+		t := time.NewTimer(to)
```

Comment:
would be better to initialize timer outside of goroutine and pass it as arg. 
can take some time to start a goroutine